### PR TITLE
feat: Add a default CSP policy to the response headers

### PR DESCRIPTION
**Pull request by netroms**
See: [[https://jira.dhis2.org/browse/DHIS2-10695](https://jira.dhis2.org/browse/DHIS2-10695)]

Signed-off-by: Morten Svanaes <msvanaes@dhis2.org>

### CODE REVIEW

**Comment by mediremi** [comment]
CSP-related issues on the dashboard:

- We need a laxer `img-src` policy that allows `data:` URIs
- We need a laxer `connect-src` policy that allows loading maps from fastly.net
- We need a laxer `script-src` policy that allows loading `blob:` URIs from the same base URL

**Comment by amcgee** [comment]
@mediremi where are the CDNs needed?

**Comment by mediremi** [comment]
@amcgee for the dashboard, fastly is used by mapbox while the capture app uses cdnjs to load [leafletjs](https://leafletjs.com).

**Comment by amcgee** [comment]
OK good to know, thanks @mediremi 

@turban would it be possible to tell Mapbox NOT to load from a CDN?

And @dhis2/team-tracker (@JoakimSM @paschalidi @Bekkalizer) we shouldn't be loading Leaflet from a CDN - it breaks offline use-cases among other things - so it would be good to bundle Leaflet with the app itself (and even better to re-use the mapping engine from Maps, as we've discussed before).

**Comment by paschalidi** [comment]
I created a ticket for this @amcgee https://jira.dhis2.org/browse/TECH-584

**Comment by mediremi** [comment]
I have tested every app on https://debug.dhis2.org/csptesting4 (which uses the CSP from [this commit](https://github.com/dhis2/dhis2-core/pull/7745/commits/1666e9244d32fa4981c2593f5399e14efcd3e19e)) and have not run into any CSP errors.

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java [hunk 1]
```diff
@@ -149,7 +149,16 @@
     AUDIT_ENABLED( "system.audit.enabled", Constants.TRUE, false ),
     TRACKER_IMPORT_PREHEAT_CACHE_ENABLED( "tracker.import.preheat.cache.enabled", Constants.ON, false ),
     ENABLE_OAUTH2_AUTHORIZATION_SERVER( "oauth2.authorization.server.enabled", Constants.ON, false ),
-    ENABLE_JWT_OIDC_TOKEN_AUTHENTICATION( "oidc.jwt.token.authentication.enabled", Constants.OFF, false );
+    ENABLE_JWT_OIDC_TOKEN_AUTHENTICATION( "oidc.jwt.token.authentication.enabled", Constants.OFF, false ),
+    CSP_NONCE_ENABLED( "csp.nonce.enabled", Constants.ON, true ),
+    CSP_UPGRADE_INSECURE_ENABLED( "csp.upgrade.insecure.enabled", Constants.OFF, true ),
+    CSP_HEADER_VALUE( "csp.header.value", "prefetch-src 'self'; "
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java:155]: Is this value getting slightly long for dhis.conf property? Could we perhaps instead "hard-code" the default values/domains somewhere else, and then have this property for extra domains only which are then appended to the default? My point is that is implementations want to add extra domains they won't have to include all this text.

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java [hunk 2]
```diff
@@ -149,7 +149,16 @@
...
+        + "script-src 'self' 'nonce-{nonce}' 'unsafe-inline'; "
+        + "connect-src 'self' *.fastly.net; img-src 'self' data: *.fastly.net; "
```
Comment (larshelge) [inline dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java:157]: Does this affect things like OpenStreetMap, Bing maps, Google Earth Engine (used in Maps)?

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java (hunk 1 above)
Comment (netroms) [reply dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java:155]: Yes, I agree this is getting too big. I can move this into a separate class.

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java (hunk 2 above)
Comment (netroms) [reply dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java:157]: Yes, we need to add this either as default (static) or maybe make it optional aka. more dynamic. It's still a lot to figure out here. I'm trying to set up a CSP multi-end taskforce for this, so we can coordinate.

**Pull request closed without merging**